### DISCUSSION TITLE
Machine independent manifest

### DISCUSSION THREAD

**Discussion by brad-intel** [gramineproject/gramine#981] [Q&A]
When creating a manifest for an application, the manifest details such as enclave size, number of threads and/or CPU instruction set cannot be determined at build time (especially when using CI/CD). That forces us to finalise the manifest before execution and sign the binaries there as well. For AI workloads that takes quite a bit of time. 
What would be the best approach to have the manifest machine independent (or capable of dynamically adjusting)?

**Comment by dimakuv** [comment]
I don't think there is much more that can be done -- what you're already doing (finalize the manifest on the machine right before execution) seems to be reasonable for CI/CD purposes.

Generally, you list three kinds of cases:
1. Enclave size -- this is indeed a serious limitation of SGX v1 design, and many people complain about it. Unfortuntately, the only solution is to move to SGX v2 hardware & software, i.e. with EDMM support. (EDMM support is available on modern CPUs, on Linux 6.0 and will soon be available in Gramine.)
2. Number of threads -- I wouldn't call it a problem. Just specify the number of threads as the max you think your workload will ever need. Yes, each thread requires maybe 20KB of enclave memory, but in general specifying `sgx.thread_num = 1024` should be enough for everything and doesn't have negative consequences.
3. CPU features -- this is a tricky one. In general, you can just keep the defaults that Gramine provides and don't bother about it. In some very specific cases, you would want to e.g. only run on machines that have Intel AMX feature -- but then the problem is not with Gramine's manifest but with your SGX requirements.

If I misunderstood some points, or you have some proposal, feel free to reply.

**Reply by brad-intel** [reply]
Finalising right before execution means the binaries that are getting run are not signed (and signature distributed). So at execution time the binaries/data are considered trusted before being loaded in the enclave. Is that a weak link?
Thanks for the suggestions about threads. Probably it would be good to describe this parameter as `max_threads` instead on `num_threads`: we were under the impression that this number needs to be close to what the actual workload uses.
CPU features is indeed manageable (not dissimilar from having a binary compiled with a specific instruction set: it is different binary )

**Reply by dimakuv** [reply]
> Finalising right before execution means the binaries that are getting run are not signed (and signature distributed). So at execution time the binaries/data are considered trusted before being loaded in the enclave. Is that a weak link?

I thought we're talking purely in the context of CI/CD *testing*. But I guess I misunderstood you. What I meant is "finalising right before execution" is a viable option for testing, but of course not for production (it's insecure). So yes, it's a weak leak in production environments, for sure.

> Probably it would be good to describe this parameter as `max_threads` instead on `num_threads`: we were under the impression that this number needs to be close to what the actual workload uses.

Good suggestion. @boryspoplawski @mkow @kailun-qin @aneessahib Is this a reasonable name? We can deprecate but still support `num_threads`, and promote `max_threads` as a new name.

**Reply by aneessahib** [reply]
`max_threads` LGTM

**Reply by boryspoplawski** [reply]
There is one argument that `num_threads` at least hints you that the whole amount is always used/consumed (e.g. we allocate exactly that many TCS slots) and `max_threads` suggest it's just an upper bound.
Exact semantics of this variable are: create that many SGX threads, but host-OS threads are allocated on demand.

**Reply by brad-intel** [reply]
> > Finalising right before execution means the binaries that are getting run are not signed (and signature distributed). So at execution time the binaries/data are considered trusted before being loaded in the enclave. Is that a weak link?
> 
> I thought we're talking purely in the context of CI/CD _testing_. But I guess I misunderstood you. What I meant is "finalising right before execution" is a viable option for testing, but of course not for production (it's insecure). So yes, it's a weak leak in production environments, for sure.

Getting back to a production environment:  what would be the best approach? Generate multiple manifests for groups of hardware capabilities (e.g. low, med, high) and generate signature for each? Would that work?

**Reply by brad-intel** [reply]
> There is one argument that `num_threads` at least hints you that the whole amount is always used/consumed (e.g. we allocate exactly that many TCS slots) and `max_threads` suggest it's just an upper bound. Exact semantics of this variable are: create that many SGX threads, but host-OS threads are allocated on demand.

We have continuously struggled with thread allocation; it felt that we have to allocate them ahead of time (and, if we did not allocate enough, the workload would fail). E.g. AI workload or Java ones, all thread hungry

**Reply by boryspoplawski** [reply]
> We have continuously struggled with thread allocation; it felt that we have to allocate them ahead of time (and, if we did not allocate enough, the workload would fail). E.g. AI workload or Java ones, all thread hungry

What do you mean by "allocation"? Just the manifest `num_threads` option?

**Reply by dimakuv** [reply]
> Getting back to a production environment: what would be the best approach? Generate multiple manifests for groups of hardware capabilities (e.g. low, med, high) and generate signature for each? Would that work?

Actually, why to have three types of HW capabilities? Why not just let the SGX enclave query the HW features that it can use?

By default, Gramine sets an attribute mask (SIGSTRUCT.ATTRIBUTEMASK) on HW features so that they can be enabled or disabled, and the SGX enclave will still run. Like, by default Gramine runs fine on both AVX-enabled machine and no-AVX machines.

This of course slightly weakens attestation claims because now the claim "SGX enclave guaranteed to run on machines with AVX enabled only", but in most (all?) cases this attestation claim is useless.

I would say that for the vast majority of users of Gramine-SGX, there is no need to know/care about HW capabilities.

**Reply by mkow** [reply]
+1 to `max_threads` (and other similar options, if we have any). @boryspoplawski: IMO it's obvious that if you have a "max_something" setting that means the software needs to preallocate some resources for them - otherwise you wouldn't need such option at all. Anyways, even without this argument I think it's still less confusing than the current name.

@brad-intel: Summarizing other's responses: Why can't you just create one manifest which works on all your machines?

**Reply by brad-intel** [reply]
> > We have continuously struggled with thread allocation; it felt that we have to allocate them ahead of time (and, if we did not allocate enough, the workload would fail). E.g. AI workload or Java ones, all thread hungry
> 
> What do you mean by "allocation"? Just the manifest `num_threads` option?

Yes

**Reply by brad-intel** [reply]
> @brad-intel: Summarizing other's responses: Why can't you just create one manifest which works on all your machines?

I would love to but that doesn't seem to work. Obviously, if I ask for a bigger enclave that will only work on such machines. If I use a smaller sized enclave the workload might not work as, in many cases, the workloads are not SGX aware.
(Let's put aside the instruction set discussion; that is not an issue). 
The misnaming of the threads parameter was one of the issues; now I know I can use that to specify the max number of threads so that takes one issue out. The enclave size is the last one; we have to adjust the size when the workload changes ( e.g. bigger AI model, more pre-allocated memory etc.). We need a way to make sure that the workload that can work on a machine outside the enclave can still work inside; we haven't been able to find a good way (ideally automated) for that.

**Reply by dimakuv** [reply]
https://github.com/gramineproject/gramine/pull/982 -- I created a PR that deprecates `thread_num` and `rpc_thread_num` in favor of `max_threads` and `rpc_max_threads`. Similar PRs will be created for `examples` and `gsc` repos.

**Reply by dimakuv** [reply]
- https://github.com/gramineproject/gsc/pull/102 -- PR for GSC.
- https://github.com/gramineproject/examples/pull/41 -- PR for Examples.

**Reply by mkow** [reply]
> If I use a smaller sized enclave the workload might not work as, in many cases, the workloads are not SGX aware.

Hmm, how do you want to have the same manifest and signatures for different workloads? In almost all cases the idea is to have the MRSIGNER representing the exact workload inside enclave.

Also, this seems to be a startup performance only problem? You can set the size to max(all workloads) and be done, at the cost of longer enclave loading times for small workloads.
